File: library/src/main/java/com/raizlabs/android/dbflow/sql/queriable/StringQuery.java
package com.raizlabs.android.dbflow.sql.queriable;

import android.database.Cursor;

import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.language.BaseModelQueriable;
import com.raizlabs.android.dbflow.structure.Model;

/**
 * Description: Provides a very basic query mechanism for strings. Allows you to easily perform custom SQL query string
 * code where this library does not provide. It only runs a
 * {@link android.database.sqlite.SQLiteDatabase#rawQuery(String, String[])}.
 */
public class StringQuery<ModelClass extends Model> extends BaseModelQueriable<ModelClass> implements Query, ModelQueriable<ModelClass> {

    /**
     * The full SQLite query to use
     */
    private final String query;

    /**
     * Creates an instance of this class
     *
     * @param table The table to use
     * @param sql   The sql statement to query the DB with. Does not work with {@link com.raizlabs.android.dbflow.sql.language.Delete},
     *              this must be done with {@link android.database.sqlite.SQLiteDatabase#execSQL(String)}
     */
    public StringQuery(Class<ModelClass> table, String sql) {
        super(table);
        query = sql;
    }

    @Override
    public String getQuery() {
        return query;
    }

    @Override
    public Cursor query() {
        return FlowManager.getDatabaseForTable(getTable()).getWritableDatabase().rawQuery(query, null);
    }

}

File: dbflow-common/src/main/kotlin/com/raizlabs/dbflow5/query/StringQuery.kt
package com.raizlabs.dbflow5.query

import android.database.sqlite.SQLiteDatabase
import com.raizlabs.dbflow5.config.FlowLog
import com.raizlabs.dbflow5.database.DatabaseStatement
import com.raizlabs.dbflow5.database.DatabaseStatementWrapper
import com.raizlabs.dbflow5.database.DatabaseWrapper
import com.raizlabs.dbflow5.database.FlowCursor
import com.raizlabs.dbflow5.runtime.NotifyDistributor
import com.raizlabs.dbflow5.sql.Query
import com.raizlabs.dbflow5.structure.ChangeAction

/**
 * Description: Provides a very basic query mechanism for strings. Allows you to easily perform custom SQL cursor string
 * code where this library does not provide. It only runs a
 * [SQLiteDatabase.rawQuery].
 */
class StringQuery<T : Any>
/**
 * Creates an instance of this class
 *
 * @param table The table to use
 * @param query   The sql statement to query the DB with. Does not work with [Delete],
 * this must be done with [DatabaseWrapper.execSQL]
 */
(table: Class<T>,
 override val query: String)
    : BaseModelQueriable<T>(table), Query, ModelQueriable<T> {
    private var args: Array<String>? = null

    override// we don't explicitly know the change, but something changed.
    val primaryAction: ChangeAction
        get() = ChangeAction.CHANGE

    override fun cursor(databaseWrapper: DatabaseWrapper): FlowCursor? = databaseWrapper.rawQuery(query, args)

    override fun queryList(databaseWrapper: DatabaseWrapper): MutableList<T> {
        FlowLog.log(FlowLog.Level.V, "Executing query: " + query)
        return listModelLoader.load(cursor(databaseWrapper), databaseWrapper)!!
    }

    override fun querySingle(databaseWrapper: DatabaseWrapper): T? {
        FlowLog.log(FlowLog.Level.V, "Executing query: " + query)
        return singleModelLoader.load(cursor(databaseWrapper), databaseWrapper)!!
    }

    override fun <QueryClass : Any> queryCustomList(queryModelClass: Class<QueryClass>,
                                                    databaseWrapper: DatabaseWrapper)
        : MutableList<QueryClass> {
        val query = query
        FlowLog.log(FlowLog.Level.V, "Executing query: " + query)
        return getListQueryModelLoader(queryModelClass)
            .load(cursor(databaseWrapper), databaseWrapper)!!
    }

    override fun <QueryClass : Any> queryCustomSingle(queryModelClass: Class<QueryClass>,
                                                      databaseWrapper: DatabaseWrapper)
        : QueryClass? {
        val query = query
        FlowLog.log(FlowLog.Level.V, "Executing query: " + query)
        return getSingleQueryModelLoader(queryModelClass)
            .load(cursor(databaseWrapper), databaseWrapper)
    }

    override fun compileStatement(databaseWrapper: DatabaseWrapper): DatabaseStatement {
        val query = query
        FlowLog.log(FlowLog.Level.V, "Compiling Query Into Statement: query = $query , args = $args")
        return DatabaseStatementWrapper(databaseWrapper.compileStatement(query, args), this)
    }

    /**
     * Set selection arguments to execute on this raw query.
     */
    fun setSelectionArgs(args: Array<String>) = apply {
        this.args = args
    }
}
